k8s operator nil pointer panic

### What is the issue?

The operator somehow got into a bad state reconciling a service that no longer exists. This resulted in a crashloop due to nil pointer panic

```
{"level":"info","ts":1676659995.057725,"logger":"controller-runtime.metrics","msg":"Metrics server is starting to listen","addr":":8080"}
{"level":"info","ts":1676659995.0579689,"logger":"startup","msg":"Startup complete, operator running"}
{"level":"info","ts":1676659995.0591836,"msg":"Starting server","path":"/metrics","kind":"metrics","addr":":8080"}
{"level":"info","ts":1676659995.05929,"msg":"Starting EventSource","controller":"service","controllerGroup":"","controllerKind":"Service","source":"kind source: *v1.Service"}
{"level":"info","ts":1676659995.059305,"msg":"Starting EventSource","controller":"service","controllerGroup":"","controllerKind":"Service","source":"kind source: *v1.StatefulSet"}
{"level":"info","ts":1676659995.0593073,"msg":"Starting EventSource","controller":"service","controllerGroup":"","controllerKind":"Service","source":"kind source: *v1.Secret"}
{"level":"info","ts":1676659995.0593088,"msg":"Starting Controller","controller":"service","controllerGroup":"","controllerKind":"Service"}
{"level":"info","ts":1676659995.1607363,"msg":"Starting workers","controller":"service","controllerGroup":"","controllerKind":"Service","worker count":1}
{"level":"info","ts":1676659995.1659865,"msg":"Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference","controller":"service","controllerGroup":"","controllerKind":"Service","Service":{"name":"my-service","namespace":"my-namespace"},"namespace":"my-namespace","name":"my-service","reconcileID":"e727b414-ad0b-4ede-9dd3-00cc6e8efd33"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x110 pc=0x15ca744]

goroutine 714 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
    sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:118 +0x19c
panic({0x17f57e0, 0x2d5e260})
    runtime/panic.go:884 +0x1f4
main.(*ServiceReconciler).getDeviceInfo(0x4000265380, {0x1e10d08, 0x400001a990}, 0x4000189400)
    tailscale.com/cmd/k8s-operator/operator.go:548 +0x1c4
main.(*ServiceReconciler).maybeCleanup(0x4000265380, {0x1e10d08, 0x400001a990}, 0x3e?, 0x4000189400)
    tailscale.com/cmd/k8s-operator/operator.go:347 +0x580
main.(*ServiceReconciler).Reconcile(0x4000265380, {0x1e10d08, 0x400001a990}, {{{0x4002a5f500?, 0x30?}, {0x4002a6aa98?, 0x1?}}})
    tailscale.com/cmd/k8s-operator/operator.go:303 +0x4cc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1e10c60?, {0x1e10d08?, 0x400001a990?}, {{{0x4002a5f500?, 0x19d1de0?}, {0x4002a6aa98?, 0xf4dad4ac64fd9bc3?}}})
    sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:121 +0x8c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0x4002a6edc0, {0x1e10c60, 0x400279fd10}, {0x1887de0?, 0x40047b3c20?})
    sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:320 +0x280
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x4002a6edc0, {0x1e10c60, 0x400279fd10})
    sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273 +0x1a0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
    sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234 +0x74
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
    sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:230 +0x43c
Stream closed EOF for tailscale/operator-59647b44cc-5968z (operator)
```

### Steps to reproduce

Unsure exactly the steps that caused this. But I believe these were the general steps

- install operator
- deploy service
- everything works fine
- cause the operator to crash (I made changes to the manifest testing it with userspace networking, unsure what the error was causing the crash)
- delete the service
- redeploy the operator with functional manifest

### Are there any recent changes that introduced the issue?

_No response_

### OS

Linux

### OS version

_No response_

### Tailscale version

unstable

### Other software

kubernetes - minikube

### Bug report

_No response_

Hm, I'm also running in to this and I'm unsure how I ended up here. 

```
kubectl logs pod/operator-7c8965986c-7g8qg -n tailscale-operator --previous
{"level":"info","ts":1682720860.3288403,"logger":"controller-runtime.metrics","msg":"Metrics server is starting to listen","addr":":8080"}
{"level":"info","ts":1682720860.3290238,"logger":"startup","msg":"Startup complete, operator running"}
{"level":"info","ts":1682720860.329271,"msg":"Starting server","path":"/metrics","kind":"metrics","addr":"[::]:8080"}
{"level":"info","ts":1682720860.3293748,"msg":"Starting EventSource","controller":"service","controllerGroup":"","controllerKind":"Service","source":"kind source: *v1.Service"}
{"level":"info","ts":1682720860.32941,"msg":"Starting EventSource","controller":"service","controllerGroup":"","controllerKind":"Service","source":"kind source: *v1.StatefulSet"}
{"level":"info","ts":1682720860.3294146,"msg":"Starting EventSource","controller":"service","controllerGroup":"","controllerKind":"Service","source":"kind source: *v1.Secret"}
{"level":"info","ts":1682720860.329418,"msg":"Starting Controller","controller":"service","controllerGroup":"","controllerKind":"Service"}
{"level":"info","ts":1682720860.4303677,"msg":"Starting workers","controller":"service","controllerGroup":"","controllerKind":"Service","worker count":1}
{"level":"info","ts":1682720860.439665,"msg":"Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference","controller":"service","controllerGroup":"","controllerKind":"Service","Service":{"name":"graphql-api","namespace":"pre-prod"},"namespace":"pre-prod","name":"graphql-api","reconcileID":"db320ce1-bd51-4135-b521-353f26406d54"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x110 pc=0x1c8a0b1]

goroutine 604 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x1ebfe60, 0x349ce40})
	runtime/panic.go:884 +0x213
main.(*ServiceReconciler).getDeviceInfo(0xc0032823c0, {0x24f0fb0, 0xc00331ab70}, 0xc00332ec80)
	tailscale.com/cmd/k8s-operator/operator.go:569 +0x1f1
main.(*ServiceReconciler).maybeCleanup(0xc0032823c0, {0x24f0fb0, 0xc00331ab70}, 0xc0000128f0?, 0xc00332ec80)
	tailscale.com/cmd/k8s-operator/operator.go:368 +0x6d2
main.(*ServiceReconciler).Reconcile(0xc0032823c0, {0x24f0fb0, 0xc00331ab70}, {{{0xc003201b78?, 0x0?}, {0xc003201b80?, 0x40e1c7?}}})
	tailscale.com/cmd/k8s-operator/operator.go:324 +0x629
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x24f0f08?, {0x24f0fb0?, 0xc00331ab70?}, {{{0xc003201b78?, 0x20a7cc0?}, {0xc003201b80?, 0xc0032dd7a8?}}})
	sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:121 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc002ae68c0, {0x24f0f08, 0xc002af6d20}, {0x1f55ea0?, 0xc003081640?})
	sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:320 +0x309
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc002ae68c0, {0x24f0f08, 0xc002af6d20})
	sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:230 +0x587
	```

I think it had to do with me editing a service, which got the operator in to a bad state. But I don't see any dangling services or any finalizers. Here's what my operator namespace looks like:

```
kubectl get all -n tailscale-operator
NAME                            READY   STATUS             RESTARTS      AGE
pod/operator-7c8965986c-7g8qg   0/1     CrashLoopBackOff   5 (54s ago)   3m53s

NAME                       READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/operator   0/1     1            0           3m54s

NAME                                  DESIRED   CURRENT   READY   AGE
replicaset.apps/operator-7c8965986c   1         1         0       3m54s
```

And here are the secrets in that namespace:

```
kubectl get secrets -n tailscale-operator
NAME                                       TYPE                                  DATA   AGE
default-token-dvj8h                        kubernetes.io/service-account-token   3      21h
operator                                   Opaque                                4      4m15s
operator-oauth                             Opaque                                2      21h
operator-token-7rlnj                       kubernetes.io/service-account-token   3      4m17s
proxies-token-7gt7d                        kubernetes.io/service-account-token   3      4m17s
sh.helm.release.v1.tailscale-operator.v1   helm.sh/release.v1                    1      4m17s
```

If I uninstall and reinstall the manifests the problem persists which suggests to me that there is a bit of state somewhere that I'm missing. Any idea what it could be?

--- a/cmd/k8s-operator/operator.go
+++ b/cmd/k8s-operator/operator.go
@@ -567,6 +567,9 @@ func (a *ServiceReconciler) getDeviceInfo(ctx context.Context, svc *corev1.Servi
 	if err != nil {
 		return "", "", err
 	}
+	if sec == nil {
+		return "", "", nil
+	}
 	id = string(sec.Data["device_id"])
 	if id == "" {
 		return "", "", nil

